server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparison.java

```diff
@@ -49,22 +48,13 @@ public QProfileComparison(DbClient dbClient) {
     this.dbClient = dbClient;
   }
 
-  public QProfileComparisonResult compare(String leftKey, String rightKey) {
-    try (DbSession dbSession = dbClient.openSession(false)) {
-      QProfileComparisonResult result = new QProfileComparisonResult();
-      compare(dbSession, leftKey, rightKey, result);
-      return result;
-    }
-  }
-
-  private void compare(DbSession session, String leftKey, String rightKey, QProfileComparisonResult result) {
-    result.left = dbClient.qualityProfileDao().selectByKey(session, leftKey);
-    Preconditions.checkArgument(result.left != null, String.format("Could not find left profile '%s'", leftKey));
-    result.right = dbClient.qualityProfileDao().selectByKey(session, rightKey);
-    Preconditions.checkArgument(result.right != null, String.format("Could not find right profile '%s'", leftKey));
+  public QProfileComparisonResult compare(DbSession dbSession, QualityProfileDto left, QualityProfileDto right) {
+    QProfileComparisonResult result = new QProfileComparisonResult();
```

Comment:
I agree. I changed left and right to be immutable, the other properties will also be final, but still be mutable. It would be great to make the `Result` immutable as a whole, some day.